### ISSUE TITLE
Not compatible with Tomcat 9

### ISSUE THREAD

**Issue by vmassol** [paultuckey/urlrewritefilter#246]
Tomcat 9 performs very early mapping of incoming requests before the URLRewriteFilter has a chance to execute (this was working fine in Tomcat 8 and probably earlier versions too).

See https://jira.xwiki.org/browse/XWIKI-19444?focusedCommentId=112420&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-112420 for an explanation based on the XWiki use case of URLRewriteFilter

**Comment by paultuckey** [comment]
UrlRewriteFilter has fallen behind in it's support of modern versions of app servers.  It think it would be best to shop around for an alternative. 

Tomcat 9 includes it's own rewrite engine.  Perhaps it would be better to use that?   
https://tomcat.apache.org/tomcat-9.0-doc/rewrite.html

Note, I'm more than happy to accept any PRs that have full test coverage across a variety of containers.

**Comment by vmassol** [comment]
Thanks for the answer @paultuckey

Yeah I had realized it has fallen behind. It's still a great and elegant solution though ;)

Actually I've reported this issue here to let users know but it would probably be more accurate to report it on the Tomcat side as I believe the way they have coded Tomcat 9 prevents any URL rewriting from working transparently.

PS: yes I have noticed the rewrite valve. I need to find some time to check if it can work for our use case but I don't think it would as we are also rewriting outbound URLs and this is done through our call to `getResponse().encodeURL(url);`, which triggers the URLRewriteFilter. I don't see how any solution that is not executing under the same webapp would work to rewrite the URLs we generate in our responses (the solution would need to parse the content to make replacements AFAICS).

Thanks!

**Comment by octafishmedia** [comment]
I've recently migrated a java legacy app running struts1 (originally developed back in 2005-2008!) but had consistent crashes (every other day) on our virtual web server running Tomcat 9. I rolled back the urlrewritefilter to version 4.0.3 and found stability. However after a few weeks on intermittent crashes (tomcat freezes when heap size maxes out) I;'m considering eliminating it all together to avoid issues.

Any recommendations for replacing this well used tool for years?

**Comment by paultuckey** [comment]
Perhaps use Nginx in front as rewriting proxy, the built-in tomcat rewrite valve mentioned above.  It's also relatively simple to write your own custom filter for your specific use-case.

**Comment by paultuckey** [comment]
`5.1.1` should now work with tomcat 9.

**Comment by octafishmedia** [comment]
Tomcat 9 recommends Servlet 4.0 usage (https://tomcat.apache.org/whichversion.html)

Updated README says to use older version 4.0.3 w/ Servlet spec 4.0

Can I use Tomcat 9 + version 5.1.2 + Servlet 4.0 ?

**Comment by paultuckey** [comment]
Apologies, I was confused about this.  Looks like Tomcat 9 users will have to stay on version 4.0.3.

If anyone want to submit a PR to fix this issue (use the 4.x branch) I'd be happy to accept it.

**Comment by stevesobol** [comment]
I'm looking for feedback on https://github.com/lobosstudios/xwiki_urlrewrite

I have it working with Tomcat 15.5.0, and am open to requests to making it work with older versions of Tomcat.

**Comment by stevesobol** [comment]
Looking for feedback on https://github.com/lobosstudios/xwiki_urlrewrite

I'm running XWiki 15.5.0/Tomcat 9.0.76 and URL rewrites seem to work fine.

**Comment by acotiuga** [comment]
@stevesobol I'm updating here for you: https://github.com/lobosstudios/xwiki_urlrewrite still presents the incompatibility reported in this issue, but only on certain actions as discussed in: https://forum.xwiki.org/t/xwiki-docker-image-with-short-urls/13448/9

**Comment by stevesobol** [comment]
@acotiuga yes, I think I misunderstood the original problem. I'll dig into it when I have time.

**Comment by paultuckey** [comment]
Closing this. I've gone through the linked XWiki investigation, this isn't
something UrlRewriteFilter can fix.

**Root cause** (credit to @vmassol's analysis in XWIKI-19444): Tomcat 9 maps a
request to a servlet very early, in `CoyoteAdapter.postParseRequest()`, using the
*original* URI - before any filter runs. If that URI matches no
`<servlet-mapping>` it goes to the DefaultServlet, which has no multipart config.
For a `multipart/form-data` POST the body then never gets parsed, so form fields
(in XWiki's case the CSRF `form_token`) come back null.

The rewrite itself works fine - XWiki's debug logs confirm the URL is rewritten as
expected. But servlet selection and the multipart-parsing decision have both
already happened by the time the filter chain is entered. Tomcat 8 and Jetty don't
behave this way, and there's no hook available to a filter to change it.

**Workarounds:**

- Preferred: add a `<servlet-mapping>` for the un-rewritten URL pattern, so
  Tomcat's early mapping lands on your real (multipart-capable) servlet.
- Or set `allowCasualMultipartParsing="true"` in `context.xml`. Note that people
  in the XWiki thread found this had to go in Tomcat's global `conf/context.xml`
  rather than the webapp's `META-INF/context.xml`, and Tomcat documents it as not
  spec-compliant.

**Separately:** several comments here are about
`NoClassDefFoundError: jakarta/servlet/Filter`. That's a different problem - 5.x
targets `jakarta.servlet` (Tomcat 10.1+); Tomcat 9 needs a javax build.

**Comment by vmassol** [comment]
@paultuckey Thx for taking the time to reply! Appreciated.
